src/Bucket.hpp

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <array>
+#include <optional>
+
+template<class T, size_t N>
+class Bucket {

```

Comment:
Do we really need the Bucket™️ though? If it's for the performance gain of allocating everything at once, `std::vector` has `reserve()`. If it's for the performance gain of allocating on the stack, `std::array` has a `size()` that is distinct from `max_size()`. Is there some other benefit that I'm not seeing? Is it for the tidy `add()` and `isFull()` helpers?

Reply:
it's for stack allocation, but see this snippet from [cppreference](https://en.cppreference.com/w/cpp/container/array/max_size):
> Because each [std::array](http://en.cppreference.com/w/cpp/container/array)<T, N> is a fixed-size container, the value returned by max_size equals N (which is also the value returned by [size()](https://en.cppreference.com/w/cpp/container/array/size)).

So they are different functions, but we still need to keep track of how many spots we have in the array, which is why I made a helper class for it